yorkie/src/main/kotlin/dev/yorkie/util/SplayTreeSet.kt

```diff
@@ -0,0 +1,364 @@
+package dev.yorkie.util
+
+import com.google.common.annotations.VisibleForTesting
+
+/**
+ * SplayTreeSet is weighted binary search tree set based on Splay tree.
+ * original paper on Splay Trees:
+ * @link https://www.cs.cmu.edu/~sleator/papers/self-adjusting.pdf
+ */
+// should SplayTree implement MutableSet?
+internal class SplayTreeSet<V>(private val lengthCalculator: (V) -> Int = { 1 }) {
+    private val valueToNodes = mutableMapOf<V, Node<V>>()
+
+    @VisibleForTesting
+    var root: Node<V>? = null
+        private set
+
+    val length
+        get() = root?.weight ?: 0
+
+    private val maximum: Node<V>?
+        get() {
+            var node = root
+            while (node?.right != null) {
+                node = node.right
+            }
+            return node
+        }
+
+    private fun subtree(node: Node<V>?): SplayTreeSet<V> {
+        return SplayTreeSet(lengthCalculator).also {
+            it.root = node
+        }
+    }
+
+    /**
+     * returns the Value and offset of the given index.
+     */
+    @Suppress("UNCHECKED_CAST")
+    fun find(pos: Int): ValueToOffset<V> {
+        var target = pos
+        val root = root
+        if (root == null || target < 0) {
+            return ValueToOffset.Empty as ValueToOffset<V>
```

Comment:
Do we need to add UNCHECKED_CAST because of `as`?

Reply:
Yes although it seems ugly, this pattern is used quite a lot for optimizations for empty or default types.
This way we can share a single instance across any generic types.